### ISSUE TITLE
Test Plan for April 2019 release

### ISSUE THREAD

**Issue by ramya-rao-a** [Microsoft/vscode-go#2410]
If you got a notification from this issue, then you are part of the next update to the Go extension which I plan to release early next week.
Thanks and Welcome!

Here, I am listing some of the features and the bug fixes that will go in the next update and for which I need some help in testing. Since I dont work with Go in my day job, I always feel that someone who does Go more than me can help in testing tremendously.

Also mentioned are the folks who logged the bugs/feature requests, those who upvoted it, and those who fixed it by sending PRs as these are the people with the most context on what the issue is.

And when I say "test", you can do a minimum pass of just installing the update and then do nothing special other than going about using VS Code like you do everyday. Or you can test the feature you requested/implemented or the bug you logged/fixed. Or you can test all of the below. Every bit helps.

Drop in a comment in this issue after 2 or 3 days of [installing the update](https://github.com/Microsoft/vscode-go/wiki/Use-the-beta-version-of-the-latest-Go-extension) i.e the beta version with your findings.

Those using modules will get a prompt to use `gopls`, the language server from Google as it performs much better especially for code completions. Please try it out.

### Go Modules support improvements

* @calebdoxsey, @brainsnail, @yookoala, @samadadi, @jclc
    * Add grammar for `go.mod` and `go.sum` files, thus providing syntax highlighting for them. [Feature Request 1886](https://github.com/Microsoft/vscode-go/issues/1886) implemented with [PR 2344](https://github.com/Microsoft/vscode-go/pull/2344)

* @FAQinghere, @nezorflame, @stevvooe
  * Use `goimports` for formatting when using Go modules without the language server because `goreturns`(the default formatting tool) doesn't work with modules yet. Fixes [Bug 2309](https://github.com/Microsoft/vscode-go/issues/2309)

* @nezorflame, @qjerome
  * Fix build on save, install and debug features when `GO111MODULE` is set to `on` inside the GOPATH. Fixes [Bug 2238](https://github.com/Microsoft/vscode-go/issues/2238) with [commit 15f571e4](https://github.com/Microsoft/vscode-go/commit/15f571e490e35a4c531aca6585aae7d07dfaae93)
  
* @stamblerre, @primalmotion, @mafredri
  * Support `gopls`, the language server from Google as the one from Sourcegraph is no longer under active development. Also because `gopls` supports Go modules. [PR 2383](https://github.com/Microsoft/vscode-go/pull/2383). Please read our [updated README on language servers](https://github.com/Microsoft/vscode-go/tree/8ea6e4708dfd141ab65b4c7eb4f71e55d098a222#go-language-server-experimental) for the latest on what we recommend.
    

### New features

* @aswinmprabhu, @conradwt, @mgood, @Hades32, @jijeshmohan, @scriptnull, @carlca, @alwindoss 
    * Refactor commands to extract functions and variables using [godoctor](https://github.com/godoctor/godoctor). [Feature Request 588](https://github.com/Microsoft/vscode-go/issues/588) implemented with [PR 2139](https://github.com/Microsoft/vscode-go/pull/2139)

### Debugging improvements

* @vladbarosan, @monkeyWie, @mxschmitt, @chadountain1
    * Fix issue of debug file not being cleared after debugging on Windows. Fixes [Bug 2265](https://github.com/Microsoft/vscode-go/issues/2265) with [PR 2332](https://github.com/Microsoft/vscode-go/pull/2332)

* @jhendrixMSFT, @tcagan 
    * Fix variable display when debugging when the name has `-` in it. Fixes [Bug 2328](https://github.com/Microsoft/vscode-go/issues/2328) with [PR 2320](https://github.com/Microsoft/vscode-go/pull/2320)

* @aschade92, @mhr3
    * You can now control whether global variables are shown or not in the variable pane while debugging by tweaking the `showGlobalVariables` property in the `go.delveConfig` setting. [Feature Request 2323](https://github.com/Microsoft/vscode-go/issues/2323) implemented with [PR 2351](https://github.com/Microsoft/vscode-go/pull/2351)

### Others

* @FiloSottile, @vladbarosan
    * Avoid prompts to re-compile Go tools when the `go.toolsGopath` is different between workspaces. [PR 1589](https://github.com/Microsoft/vscode-go/pull/1589)

* @lggomez, @cassiobock
    * Avoid moving to the next parameter in the Signature Help feature, when provided parameter value is a string with commas. Fixes [Bug 1682](https://github.com/Microsoft/vscode-go/issues/1682) with [PR 1738](https://github.com/Microsoft/vscode-go/pull/1738)
 

* @oneslash, @norilt 
    * Fix auto-completion when there is a string with `//` in the same line before the current position. Fixes [Bug 2240](https://github.com/Microsoft/vscode-go/issues/2240) with [PR 2316](https://github.com/Microsoft/vscode-go/pull/2316)

* @OneOfOne 
    * Clean up the temporary directory created by the Go extension when VS Code window is closed. Fixes [Bug 2188](https://github.com/Microsoft/vscode-go/issues/2188) with [commit 4a241f80](https://github.com/Microsoft/vscode-go/commit/4a241f806809465afa7085a3f62729ff37fa63cd)

* @EmpireJones
    * Show the start of on save features in the output panel. Fixes [Bug 1869](https://github.com/Microsoft/vscode-go/issues/1869) with [commit 058eccf17](https://github.com/Microsoft/vscode-go/commit/058eccf17f1b0eebd607581591828531d768b98e)

* @elvizlai
    * Ignore `GOBIN` when user has set `go.toolsGopath` setting. Fixes [Bug 2339](https://github.com/Microsoft/vscode-go/issues/2339) with [commit 9f99c30](https://github.com/Microsoft/vscode-go/commit/9f99c306e00209a221abc1962c4f419565141ffb)

**Comment by oneslash** [comment]
fix for #2240 works fine for me

**Comment by mafredri** [comment]
This might actually be a `gopls` issue, I haven't investigated it further, but sometimes invalid code can cause code to be erased. It's as if we're trying to create a valid Go expression out of it.

Reporting it now as it can lead to a pretty poor user experience if released as-is.

Example (see trailing comma, not yet saved):

<img width="366" alt="image" src="https://user-images.githubusercontent.com/147409/55464210-3f653a00-5603-11e9-90f1-1339046d2b74.png">

Save:

<img width="351" alt="image" src="https://user-images.githubusercontent.com/147409/55464232-4ee48300-5603-11e9-9f4a-9640859dcf18.png">

**Comment by stamblerre** [comment]
@mafredri: this is a gopls bug, if you want to open it there. 2 questions though - could you share the full contents of the file (I'm not able to reproduce), and what does gofmt do to this file?

**Comment by nezorflame** [comment]
@ramya-rao-a you didn't attach the released beta extension :)
I've built myself one from the `0.9.3-beta.7` tag, sharing it here in case someone needs it: [Go-0.9.3-beta.7.zip](https://github.com/Microsoft/vscode-go/files/3039665/Go-0.9.3-beta.7.zip)

**Comment by nezorflame** [comment]
I can confirm that #2238 has been fixed.

**Comment by primalmotion** [comment]
With gopls, hovering on a structure shows the attributes and all on one line instead of being nicely formatted like before.

**Comment by nezorflame** [comment]
@primalmotion there's a fork of `gopls` by @saibing which has ported hover and some more functions from `bingo`, you might want to check it out: https://github.com/saibing/tools

**Comment by mafredri** [comment]
@stamblerre I can reproduce it, for example, by adding a comma at the end of this file/line: [js/net/types_js.go#L11](https://github.com/mafredri/goodspeaker/blob/e81c46afafbe13d8f6dd9606b6c133fae0f89ba3/js/net/types_js.go#L11). It deletes `tmp.Release()`.

When adding the comma to the file outside of vscode, and running `gofmt`, I see this error:

```
types_js.go:11:2: expected 1 expression
```

`gofmt -w` does not modify the file. I'll gather a trace and open a report on the Go issue tracker soon(ish).

**Comment by ramya-rao-a** [comment]
> I've built myself one from the 0.9.3-beta.7 tag, sharing it here in case someone needs it

@nezorflame Thanks! I just realized the same. I clicked on the "Publish Release" button before the upload was complete :(. The release has been updated with the vsix file now.

@stamblerre Are there issues with a particular label somewhere where I can point folks to for known issues or features that are being tracked at? A few things I see are
- Hover info, signature help, auto-completion should show docs as well
- Hover info not formatted as expected
- Auto-completion doesn't respect the setting `go.useCodeSnippetsOnFunctionSuggest` (This might need some work on my end to plumb through the settings)
- Missing imports don't get added on file save (@stamblerre, I recall you mentioned that gopls has a on save hook where you run something similar to goimports to find missing imports and add them)

**Comment by lucashtc** [comment]
Using go mod here got incredibly fast.

**Comment by FiloSottile** [comment]
Formatting seems to be broken when working on the standard library. It works correctly for my other projects. When working on the standard library, even forcing a "Format Document" action does nothing. Nothing appears in the developer console.

This is my entire workspace settings.json, so I don't think it's caused by my wrokspace.

```
{
    "go.goroot": "~/go",
    "go.toolsGopath": "~/.vscode/godev",
    "spellright.documentTypes": [
        "git-commit"
    ],
    "[html]": {
        "editor.formatOnSave": false
    }
}
```

**Comment by mxschmitt** [comment]
Now the debug file will be deleted after I stop debugging. Good Job! @ramya-rao-a

**Comment by FiloSottile** [comment]
Turns out https://github.com/Microsoft/vscode-go/issues/2410#issuecomment-479595297 is a known gopls issue.

**Comment by stamblerre** [comment]
@ramya-rao-a: All of the gopls issues can be found here: https://github.com/golang/go/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+label%3Agopls. The only one I'm not sure of how to handle is getting the settings for function placeholders, but we can discuss that offline.

@FiloSottile: I actually just checked, but I was wrong earlier - gopls should work fine in the Go standard library. Looking at your settings, it doesn't seem like you have gopls enabled? Also getting imports on save with gopls does require the additional setting of:

```
"editor.codeActionsOnSave": {
    "source.organizeImports": true
},
```

**Comment by kidlj** [comment]
Hi, @ramya-rao-a 

Here I have a  problem of vscode-go-beat consuming too much memory.

I have a fresh new vscode-insiders with no custom settings. With only vscode-go-beta activated, the editor consumed about ~10G memory, and the log says `[warning] UNRESPONSIVE extension host, 'ms-vscode.Go' took 98% of 5664.527ms`.

Environment:
OS: macOS 10.14.4 
Code version: 1.33.0-insider (1.33.0-insider)
vscode-go version: Go-0.9.3-beta.7.vsix
vscodevim version: 1.3.0

<img width="800" alt="code_memory_10g" src="https://user-images.githubusercontent.com/300616/55527545-1d83ba00-56cc-11e9-9ab8-3b519dddb1c4.png">

<img width="1317" alt="extension_log" src="https://user-images.githubusercontent.com/300616/55527552-22486e00-56cc-11e9-8d33-8b7e8f21f85f.png">

**Comment by nezorflame** [comment]
@ramya-rao-a it seems like the `goreturns` is still being preferred instead of `goimports` here for me in spite of change from #2309.
Removing `"go.formatTool": "goimports"` instantly makes the editor ask me to install `goreturns`.
My setup:
- VSCode `1.32.3`
- vscode-go `0.9.3-beta.7`
- `GO111MODULE=on` is globally set in my ENV, so modules are enabled

**Comment by primalmotion** [comment]
I also have an issue with package completion. The choices are ... weird

<img width="608" alt="Screen Shot 2019-04-04 at 10 51 04" src="https://user-images.githubusercontent.com/208711/55577121-a815ea80-56c7-11e9-98bf-9048a3a1e43b.png">

It doesn't even show the standard `json` package. Is there a map overwriting itself somewhere?

**Comment by ahmetb** [comment]
This solves my intellisense slowness with Go modules I reported at #2366.

**Comment by ahmetb** [comment]
I found a problem that some of my code gets deleted when I hit save (maybe gofmt issue?)

![image](https://user-images.githubusercontent.com/159209/55578400-a994e200-56ca-11e9-9cd8-3987ddf39b0c.png)

when my cursor is on `context.`, I hit save and  I get a few lines of code deleted (for some reason, the `u` remains):

![image](https://user-images.githubusercontent.com/159209/55578429-bc0f1b80-56ca-11e9-897b-a3d2c7450115.png)

**Comment by shreyu86** [comment]
Environment: 
OS: `macOS 10.14.4`
VSCode:  `1.32.3`
Go: `go1.12.1 darwin/amd64`
 
I found an issue when using `gopls` along with modules. I have all my dependencies resolved file but see this in the editor 
![image](https://user-images.githubusercontent.com/311217/55578855-c2ea5e00-56cb-11e9-9df6-a49503ec346c.png)

![image](https://user-images.githubusercontent.com/311217/55578886-d09fe380-56cb-11e9-92c9-f9cfecfdee72.png)

Here is output from `gopls`
![image](https://user-images.githubusercontent.com/311217/55578935-eb725800-56cb-11e9-8512-0d7ea110b22b.png)

**Comment by stamblerre** [comment]
For all of the people who are finding issues with gopls, it'd be really helpful if you could file them at [golang.org/issue](https://golang.org/issue). It will be a lot easier for me to diagnose and resolve them that way.

Since gopls is still opt-in (even though it's the default language server, it's still an experimental setting), I don't think that all of the issues with gopls should necessarily be considered blockers, but I will do my best to resolve them ASAP.

**Comment by conradwt** [comment]
@shreyu86 Do you have a test case (i.e. runnable code sample) which exhibits this issue?

**Comment by ramya-rao-a** [comment]
@kidlj Please log a new issue with your go related setting changes and details from the Process explorer  by running the command `Developer -> Open Process Explorer`

@nezorflame If you have removed the setting `"go.formatTool": "goimports"`, then for the remainder of that VS Code session, the extension will assume that you wanted `goreturns` and so the prompt when it attempts to format. When reloaded or a new window is opened for a project using modules, you will be notified about the switch back to goimports. We can look into showing a notification when user chooses anything other than `gofmt` or `goimports` when using modules asking them not to do so. Please log a new issue for this.

@primalmotion Can you log a new issue for the issue with package completion? We can then see if those are coming from the extension or gopls and handle it appropriately

@shreyu86 I don't think that is from `gopls`. Diagnostics from the language server are not enabled by default. Please log a new issue.

**Comment by nezorflame** [comment]
@ramya-rao-a actually the warning is still there even after the VSCode restart:
<img width="470" alt="изображение" src="https://user-images.githubusercontent.com/3906766/55617656-4c1d9580-579d-11e9-92c7-3a1177d2a7f9.png">
I'll file a new issue.

**Comment by lggomez** [comment]
I can confirm #1682 fix

**Comment by calebdoxsey** [comment]
I've seen no issues with "Feature Request 1886" (syntax highlighting for go.mod/go.sum).

**Comment by shreyu86** [comment]
@ramya-rao-a the errors I was seeing went away when I switched off diagnostics
```
"go.languageServerExperimentalFeatures": {
    "diagnostics": false
  },
```
I can open a new issue if this is not being tracked.

**Comment by stamblerre** [comment]
@shreyu86: I don't think those errors were from gopls, like Ramya said. If you notice, on the right side of them, they say the source, which in the case of gopls, would be "LSP".

**Comment by shreyu86** [comment]
@stamblerre Thanks for the response, https://github.com/Microsoft/vscode-go/issues/2410#issuecomment-480402003 made the errors I was seeing go away.

**Comment by monkeyWie** [comment]
@ramya-rao-a 
The issue #2265 has been fixed. Thanks!

**Comment by jclc** [comment]
`go.mod` syntax highlighting doesn't work for comments right after some tokens unless there's a space in between. Seems to work for gofmt'd files. Doesn't provide paren completion like editing .go files would.

**Comment by brainsnail** [comment]
It looks like @jclc is right - these are the following scenarios where the comment highlighting doesn't work as expected without a space:

**Doesn't recognize the comment**
* After the `go` keyword and version
* After the `module` keyword and module 
<img width="244" alt="image" src="https://user-images.githubusercontent.com/444193/55674293-c4e92400-5867-11e9-984c-10317b609fad.png">

**Recognizes the comment string, but breaks subsequent highlighting**
* After the `replace` keyword before a paren
<img width="1002" alt="image" src="https://user-images.githubusercontent.com/444193/55674298-d8948a80-5867-11e9-810b-2d496298505f.png">

Are there any other areas where you encountered this?


The paren autocompletion is also something I verified. I think this is driven by a setting - maybe we haven't flagged the .mod file as being applicable for autocomplete? Isn't the parenthesis driven by a snippet? That being said, I think that most users would be using the Go toolchain to generate their `.mod` files and wouldn't be manually updating them.

**Comment by ramya-rao-a** [comment]
@brainsnail Thanks for those screenshots, can you log a new issue for that?

@jclc, @brainsnail  We will not be providing editing support for go.mod files, and so no auto-completion

**Comment by brainsnail** [comment]
@ramya-rao-a Yep, thanks Ramya!

**Comment by treeder** [comment]
Could the import fix for gopls be turned on by default?

```json
"editor.codeActionsOnSave": {
    "source.organizeImports": true
},
```

**Comment by kyleferguson** [comment]
Auto imports and auto complete for unimported seem to be broken now when working with modules. If I turn off the use language server setting, it seems to fix it but then goes back to being super slow.

**Comment by nezorflame** [comment]
@kyleferguson you could also use `Go: Add import` command from Palette to add required import. It works pretty well.

**Comment by kyleferguson** [comment]
Yea this does work. The auto import is really handy when working with large libraries though, like Kubernetes where you may have many definitions named the same thing. Auto import would correctly guess which one base on signatures etc.

**Comment by nezorflame** [comment]
Agreed, this needs to work.

**Comment by jclc** [comment]
I'm running into this issue while running the beta:

When editing a file with the following contents, gopls repeatedly panics until the extension refuses to restart it.
```
package vulkan

func (r *Renderer) freeResources
```
gopls logs:
https://gist.github.com/jclc/fe0cafdc2682356e7f277e4ba10aa9a6

**Comment by ramya-rao-a** [comment]
Thanks for all the help everyone, the next version of the extension has been released (0.10.0)

- If you see any issue with gopls, please check existing issues at https://github.com/golang/go/issues?q=is%3Aissue+is%3Aopen+label%3Agopls before logging a new one at https://github.com/golang/go/issues
- If you see any other problems, please log a new issue in this repo
